### PR TITLE
Add vectorization to improve CRC32 performance

### PR DESCRIPTION
**Pull request by brantburnett**
This significantly improves performance for System.IO.Hashing.Crc32 for cases where the source span is 64 bytes or larger on Intel x86/x64 and modern ARM architectures. It also improves the performance on ARM in cases where vectorization is not an option, such as systems without the necessary intrinsic or for short source spans.

The vectorization change only applies to .NET 7 and later targets of System.IO.Hashing because it uses some Vector128 APIs added in .NET 7. The scalar processing ARM changes also apply to .NET 6 and later.

The vectorization algorithm is a C# implementation of the algorithm put forth in the Intel paper "Fast CRC Computation for Generic Polynomials Using PCLMULQDQ Instruction" in December 2009. It is a modernization of the implementation found in ImageSharp offered here: https://github.com/dotnet/runtime/issues/40244#issuecomment-667723216.

BenchmarkDotNet=v0.13.2.2052-nightly, OS=Windows 11 (10.0.22621.1413)
Intel Core i7-10850H CPU 2.70GHz, 1 CPU, 12 logical and 6 physical cores
.NET SDK=8.0.100-preview.1.23115.2
  [Host]     : .NET 8.0.0 (8.0.23.11008), X64 RyuJIT AVX2
  Job-UHMIUW : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-IZYDKJ : .NET 8.0.0 (42.42.42.42424), X64 RyuJIT AVX2

PowerPlanMode=00000000-0000-0000-0000-000000000000  IterationTime=250.0000 ms  MaxIterationCount=20
MinIterationCount=15  WarmupCount=1

| Method |        Job | BufferSize |        Mean |     Error |    StdDev |      Median |         Min |         Max | Ratio | RatioSD |
|------- |----------- |----------- |------------:|----------:|----------:|------------:|------------:|------------:|------:|--------:|
| Append | Current    |         16 |    31.41 ns |  0.637 ns |  0.734 ns |    31.35 ns |    30.23 ns |    32.36 ns |  1.00 |    0.00 |
| Append | Intrinsics |         16 |    30.97 ns |  0.778 ns |  0.896 ns |    31.14 ns |    29.59 ns |    32.12 ns |  0.99 |    0.04 |
|        |            |            |             |           |           |             |             |             |       |         |
| Append | Current    |        128 |   250.96 ns |  4.743 ns |  5.075 ns |   251.44 ns |   242.05 ns |   257.93 ns |  1.00 |    0.00 |
| Append | Intrinsics |        128 |    19.05 ns |  0.297 ns |  0.263 ns |    18.99 ns |    18.55 ns |    19.57 ns |  0.08 |    0.00 |
|        |            |            |             |           |           |             |             |             |       |         |
| Append | Current    |       1024 | 1,990.18 ns | 31.113 ns | 29.104 ns | 1,994.39 ns | 1,948.98 ns | 2,030.06 ns |  1.00 |    0.00 |
| Append | Intrinsics |       1024 |    58.31 ns |  1.452 ns |  1.672 ns |    58.49 ns |    55.71 ns |    60.18 ns |  0.03 |    0.00 |

BenchmarkDotNet=v0.13.2.2052-nightly, OS=ubuntu 22.04
AWS m6g.xlarge Graviton2
.NET SDK=8.0.100-preview.1.23115.2
  [Host]     : .NET 8.0.0 (8.0.23.11008), Arm64 RyuJIT AdvSIMD
  Job-LINWAX : .NET 8.0.0 (42.42.42.42424), Arm64 RyuJIT AdvSIMD
  Job-SOJHQU : .NET 8.0.0 (42.42.42.42424), Arm64 RyuJIT AdvSIMD

PowerPlanMode=00000000-0000-0000-0000-000000000000  IterationTime=250.0000 ms  MaxIterationCount=20
MinIterationCount=15  WarmupCount=1

| Method |        Job |  BufferSize |         Mean |     Error |    StdDev |       Median |          Min |          Max | Ratio |
|------- |----------- |------------ |-------------:|----------:|----------:|-------------:|-------------:|-------------:|------:|
| Append | Current    |          16 |    44.797 ns | 0.0125 ns | 0.0117 ns |    44.796 ns |    44.777 ns |    44.815 ns |  1.00 |
| Append | Intrinsics |          16 |     8.521 ns | 0.0502 ns | 0.0445 ns |     8.525 ns |     8.444 ns |     8.590 ns |  0.19 |
|        |            |             |              |           |           |              |              |              |       |
| Append | Current    |         128 |   363.017 ns | 0.0252 ns | 0.0223 ns |   363.021 ns |   362.977 ns |   363.048 ns |  1.00 |
| Append | Intrinsics |         128 |    29.491 ns | 0.0412 ns | 0.0385 ns |    29.498 ns |    29.414 ns |    29.543 ns |  0.08 |
|        |            |             |              |           |           |              |              |              |       |
| Append | Current    |        1024 | 2,887.236 ns | 0.3149 ns | 0.2946 ns | 2,887.090 ns | 2,886.898 ns | 2,887.818 ns |  1.00 |
| Append | Intrinsics |        1024 |    92.073 ns | 0.4069 ns | 0.3807 ns |    92.078 ns |    91.529 ns |    92.833 ns |  0.03 |

### CODE REVIEW

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs
Comment (bencyoung-Fignum) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs:177]: This seems to fall through to the other implementation anyway?

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (gfoidl) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs:53]: Similar to suggestion above. Work on `ref` instead of the ROS and use `LoadUnsafe` w/ the offset,
So to this method there could be passed the `ref` + `length` of the ROS, then also no `ref ROS` is needed as argument.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (gfoidl) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs:55]:
```suggestion
                x1 ^= x5;
```
Roughly a matter of taste, but I'd write it this way (other places as well).

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs
Comment (gfoidl) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs:172]: "Deconstruct" the `source` here into
```c#
ref byte src = ref MemoryMarshal.GetReference(source);
nint length = source.Length;
```
and use these values throughout the rest of the method.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs]: Done for all XOR and AND operations

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs]: Done for all XOR and AND operations

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs]: Done

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs]: I've completed this internally within the X86 method, but I'm still looking into the best option for moving away from `ref ROS` as a parameter. This method does advance the span and needs to return that to the caller so the caller can process remaining bytes dangling on the end. `ref ROS` is probably the cleanest way since you can't pass `ref ref byte` nor return `ref byte` in a `ValueTuple`.

Returning a private `ref struct` with a `ref` field is an option, but I'm not sure how well JIT will optimize that, `ref ROS` may be better. Another option may be to not take fall through to process the remaining bytes at all and process the remaining bytes directly in the X86 method.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs
Comment (jkotas) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs:177]: This will make `source` address-taken that is likely to regress performance of the loop below.

Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs]: I've rewritten this to avoid passing source by reference

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (stephentoub) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs:22]: This results in ~800 bytes of asm.  We don't want to inline it :)

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (stephentoub) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs:136]:
```suggestion
            uint result = x1.AsUInt32().GetElement(1);
            return length > 0
                ? UpdateSlowPath(result, MemoryMarshal.CreateReadOnlySpan(ref srcRef, length))
                : result;
```

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs
Comment (stephentoub) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs:188]: Let's call this UpdateScalar rather than UpdateSlowPath.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs
Comment (stephentoub) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs:187]: I don't think we want this on this method, either.  You could move it to the Update method if you were concerned about Update not getting inlined, though I expect it will, but I think having it here could end up forcing this whole method to be inlined into user call sites.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (stephentoub) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs:28]: Let's move this check to the caller, so that we're not paying for the extra method invocation in this case, and then just assert here that the length is sufficient.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (stephentoub) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs:21]: It'd be nice to include the name of the paper this is based on.

File: src/libraries/System.IO.Hashing/src/System.IO.Hashing.csproj
Comment (stephentoub) [inline src/libraries/System.IO.Hashing/src/System.IO.Hashing.csproj:16]:
```suggestion
    <Compile Include="System\IO\Hashing\Crc32.X86.cs" Condition="$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0'))" />
```

**Review by stephentoub — COMMENTED** [review]
Nice improvements. Thanks.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs]: After considering it, I agree. The constant was a holdover from a previous iteration where I was checking the length before calling the method, which made it harder to intuit the value. Since you requested that the length check be moved to the `Update` method, I've left the constant for that purpose only and renamed it appropriately. All the other sites use `Vector128<byte>.Count`.

Let me know if you still think the `Update` method should just use `Vector128<byte>.Count * 4`. I just thought it made things clearer when the logic is split between two files.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs]: Done

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs]: Done

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs]: Done

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs]: Done

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs]: Agreed, especially since we're not aggressively inlining anymore.

File: src/libraries/System.IO.Hashing/src/System.IO.Hashing.csproj
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System.IO.Hashing.csproj]: Hey, I learn a new trick every day. Thanks.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs:22]: Done

**Comment by danmoseley** [comment]
Given this references other work do we need to add an entry in the third party notices file?

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (tannergooding) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs]: If you abstract this out into a helper, you can also support it on Arm64 by using the "polynomial multiply" instructions. Believe you specifically want `PMULL` and `PMULL2` which correspond to `AdvSimd.PolynmoialMultiplyWideningLower/Upper`

> Polynomial Multiply Long. This instruction multiplies corresponding elements in the lower or upper half of the
vectors of the two source SIMD&FP registers, places the results in a vector, and writes the vector to the destination
SIMD&FP register. The destination vector elements are twice as long as the elements that are multiplied.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs
Comment (tannergooding) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs]: You only need `Pclmulqdq.IsSupported` as it implies `Sse2` support and therefore also implies `Vector128.IsHardwareAccelerated`

Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs]: Done

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs]: I actually considered using PMULL and PMULL2 on ARM when I was writing this, but I didn't do so because:

- Some of this implementation feels like it might be affected by byte order. Probably addressable, but it was a concern.
- I couldn't find an equivalent of `Sse2.ShiftRightLogical128BitLane` on ARM, so we'd need to create a less performant equivalent (correct me if I'm just blind on this one)
- ARM has a built-in CRC32 intrinsic that uses the same polynomial, so I assumed (potentially incorrectly) that it would be a better choice. So I was thinking I'd come back next and add an ARM-specific implementation.

**Comment by brantburnett** [comment]
> Given this references other work do we need to add an entry in the third party notices file?

Good question, I was hoping someone would tell me. I'm not really an expert on licensing. This work was based on the ImageSharp implementation of the algorithm, but it's a pretty major overhaul of the C# so I'd assume that any required attribution would just be to the Intel paper that it was originally based on. The Intel paper doesn't really have clear licensing, at least to my untrained eye. https://www.intel.com/content/dam/www/public/us/en/documents/white-papers/fast-crc-computation-generic-polynomials-pclmulqdq-paper.pdf

**Comment by stephentoub** [comment]
If it's a derivative of the ImageSharp code, the relevant information should be added to the third party notices file.  If it's instead just based on the algorithm in the paper, I think including the paper title in the source is sufficient.  But @richlander would have the final say here.

**Comment by brantburnett** [comment]
> If it's a derivative of the ImageSharp code, the relevant information should be added to the third party notices file. If it's instead just based on the algorithm in the paper, I think including the paper title in the source is sufficient. But @richlander would have the final say here.

For reference here is the ImageSharp code: https://github.com/SixLabors/ImageSharp/blob/f4f689ce67ecbcc35cebddba5aacb603e6d1068a/src/ImageSharp/Formats/Png/Zlib/Crc32.cs#L80

**Comment by richlander** [comment]
Overhaul or not overhaul isn't the bar nor is the bar high. If you started with the ImageSharp code, then attribution should be given. We're not trying to limit attributions but given them where warranted. From what I read, it is. Please correct me if I've got that wrong.

ImageSharp is going through a license change, but this file says Apache 2, so we're fine. /cc @jimbobsquarepants.

We should double check with our legal staff on the Intel paper. I suspect it is fine, however the licensing aspects at the end of the doc are confusing. I'll ask.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs]: I've done a bit more research on this, and it appears that the PMULL approach is more performant on modern ARM that supports it than the CRC32 intrinsic on larger buffers because it can operate on wider data sets. I see evidence of commits in the Linux kernel and Java based on this. It seems like they use the CRC32 intrinsic as a fallback when PMULL is not available.

Based on this, I'm considering refactoring this to support ARM PMULL as well. However, I'm a bit stumped on how to test it since my dev laptop is Intel. Are there any tricks documented to test using QEMU or something similar?

Comment (tannergooding) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs]: The easiest way is likely just to let our CI cover it if you don't have your own box.

I wouldn't expect any significant differences and you can filter on `BitConverter.IsLittleEndian` to ensure it works on BigEndian platforms if that's a concern.

**Comment by JimBobSquarePants** [comment]
All good by me. Would be great if someone could backport the ARM instrinics to ImageSharp though.

**Comment by richlander** [comment]
Will this change be suitable for you to use @JimBobSquarePants, or do you mean so that you have an improved implementation generally (since this change is .NET 8+)?

**Comment by brantburnett** [comment]
> All good by me. Would be great if someone could backport the ARM instrinics to ImageSharp though.

Assuming I get the ARM intrinsics working, you should be able to take System.IO.Hashing 8.0.0 (once released) as a dependency for net7.0 and forward targets to get the intrinsics, without the need to port it. If you want net6.0 support it would require a port, though. This version is currently using some APIs added in .NET 7 to Vector128.

**Comment by JimBobSquarePants** [comment]
@richlander @brantburnett 

I've just realised we've already added ARM support since that commit so no need for changes there. 

The only thing we're missing is the endiness check
https://github.com/dotnet/runtime/pull/83321/files#diff-8c4ea1d8b9624f9e2f25b7ffc2f776d6ea77492d9758f116bf1058c6882c2481R176-R180

We'd definitely delete our code once ImageSharp targets .NET 8 and use this.

**Comment by brantburnett** [comment]
I've got ARM working now and I've updated the PR title/description and benchmarks to reflect the inclusion of ARM support. I also added some additional optimizations for ARM on short source spans using the CRC32 intrinsic.

Outstanding questions I have:

- What, if any, attributions are required in the 3rd party notices file.
- Are there any changes I need to make to CI configuration to ensure System.IO.Hashing is well-tested around these intrinsics and the fallbacks?

**Comment by richlander** [comment]
I'd suggest adding an attribution for ImageSharp. We're waiting on feedback on the Intel doc.

**Comment by brantburnett** [comment]
> I'd suggest adding an attribution for ImageSharp. We're waiting on feedback on the Intel doc.

If it helps, I found this reference implementation in assembly from Intel which has a clear license: https://github.com/intel/isa-l/blob/master/crc/crc32_ieee_by4.asm 

Would that make this already covered based on this? https://github.com/dotnet/runtime/blob/main/THIRD-PARTY-NOTICES.TXT#L136 Or are further changes still required?

Sorry, I'm really not very knowledgeable on this subject.

**Comment by danmoseley** [comment]
Not sure what that existing entry is for specifically. So I’d suggest adding the license block from the top of that CRC file into that TPN file, with a note where it’s from and what it’s for. It will do no harm to have both.

**Comment by richlander** [comment]
+1

Let's go with that. I'll raise this with the folks i'm talking to, but it seems promising. I'm assuming that, as the developer of this implementation, that you find that this .asm file you linked to has the same/sufficient information as the paper such that you would have been able to move forward if you'd only had this file initially. Is that fair?

In the TPN entry, please link to the file. I believe we've done that for some other entries.

**Comment by brantburnett** [comment]
> Let's go with that. I'll raise this with the folks i'm talking to, but it seems promising. I'm assuming that, as the developer of this implementation, that you find that this .asm file you linked to has the same/sufficient information as the paper such that you would have been able to move forward if you'd only had this file initially. Is that fair?
> 
> In the TPN entry, please link to the file. I believe we've done that for some other entries.

Done, I think. Let me know if I didn't do something right. Also, yes the source code is sufficient to derive the entire algorithm without the article itself.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.Vectorized.cs
Comment (gfoidl) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.Vectorized.cs:180]: Should this be extracted to a (local) helper-function?
* with proper funciton name it's self-documenting
* the repeation from L171-L173 can be removed

**Review by gfoidl — COMMENTED** [review]
One question / suggestion, otherwise LGTM.

**Comment by richlander** [comment]
>  I'm assuming that, as the developer of this implementation, that you find that this .asm file you linked to has the same/sufficient information as the paper such that you would have been able to move forward if you'd only had this file initially

I heard back. I was told that this logic is good. If you believe you could have made this implementation with just the .asm file, then we can go forward with making a 3PN entry for just it. We should still add one for ImageSharp if you also based your implementation on that code (even if you ended up with something quite different).

**Review by gfoidl — APPROVED** [review]

**Comment by tannergooding** [comment]
Assigned myself to give a final review pass and merge if everything looks good.

Please let me know once you've resolved or responded to all feedback and I'll get started (there are a number of open, but likely already handled comments still above).

**Comment by brantburnett** [comment]
> Assigned myself to give a final review pass and merge if everything looks good.
> 
> Please let me know once you've resolved or responded to all feedback and I'll get started (there are a number of open, but likely already handled comments still above).

To my knowledge, all feedback is resolved above. I just wasn't sure what the procedure is, if I'm supposed to mark it resolved or let the reviewer mark it as resolved. I'm happy to do so if that's the procedure.

**Comment by tannergooding** [comment]
It varies from repo to repo and reviewer to reviewer, unfortunately.

The "safest" thing to do is to at least leave a little comment indicating "Fixed" or "Resolved" if its been explicitly addressed. You can optionally leave a link or comment elaborating if appropriate.

That at least lets other reviewers see that something isn't still pending or simply missed.

**Comment by brantburnett** [comment]
> It varies from repo to repo and reviewer to reviewer, unfortunately.
> 
> The "safest" thing to do is to at least leave a little comment indicating "Fixed" or "Resolved" if its been explicitly addressed. You can optionally leave a link or comment elaborating if appropriate.
> 
> That at least lets other reviewers see that something isn't still pending or simply missed.

Okay, I resolved some from gfoidl since he approved the PR and replied on the other conversations. Thanks.

**Comment by stephentoub** [comment]
Thanks. Let's get this one merged and then do the other.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.Arm.cs
Comment (tannergooding) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.Arm.cs:19]: This should really use `ReadUnaligned` since there is no guarantee that things are "properly aligned" otherwise.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.Vectorized.cs
Comment (tannergooding) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.Vectorized.cs:31]: We have a new `System.Diagnostics.UnreachableException` which is probably better.

You'd then want:
```
ThrowHelper.ThrowUnreachableException();
return default;
```

**Review by tannergooding — APPROVED** [review]

**Comment by tannergooding** [comment]
Changes overall LGTM. Should probably have a secondary review/sign-off before merging.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.Arm.cs
Comment (brantburnett) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.Arm.cs]: Rewritten to use ref byte and Unsafe.ReadUnaligned

**Pull request merged**
